api_handlers/zotero.php

```diff
@@ -25,6 +25,49 @@ function query_url_api($ids, $templates) {
        if ($template->has('ssrn'))      expand_by_zotero($template, 'https://papers.ssrn.com/sol3/papers.cfm?abstract_id=' . $template->get('ssrn'));
        if ($template->has('doi') && !doi_active($template->get('doi')))  expand_by_zotero($template, 'https://dx.doi.org/' . urlencode($template->get('doi'))); // Non-crossref DOIs, such as 10.13140/RG.2.1.1002.9609
   }
+   
+  // Now that we have expanded URLs, try to lose them
+  $ch = curl_init();
+  curl_setopt($ch, CURLOPT_FOLLOWLOCATION, TRUE);
+  curl_setopt($ch, CURLOPT_MAXREDIRS, 100); // No infinite loops for us
+  curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 4); 
+  curl_setopt($ch, CURLOPT_TIMEOUT, 15);
+  curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
+  foreach ($templates as $template) {
+    $doi = $template->get_without_comments_and_placeholders('doi');
+    $url = $template->get('url');
+    if ($doi &&
+        $url &&
+        !$template->incomplete() &&
+        !preg_match(REGEXP_DOI_ISSN_ONLY, $doi) &&
+        str_ireplace(CANONICAL_PUBLISHER_URLS, '', $url) != $url &&
+        str_ireplace(['pdf', 'image', 'plate', 'figure', 'picture'], '', $url) == $url &&
+        $template->blank(DOI_BROKEN_ALIASES))
+    {
+          curl_setopt($ch, CURLOPT_URL, "https://dx.doi.org/" . urlencode($doi));
+          if (@curl_exec($ch)) {
+            $redirectedUrl_doi = curl_getinfo($ch, CURLINFO_EFFECTIVE_URL);  // Final URL
+            $redirectedUrl_doi = strtok($redirectedUrl_doi, '?#');  // Remove session stuff
+            $url_short         = strtok($url,               '?#');
```

Comment:
```suggestion
            $url_short         = strtok($url,               '?#');
            if ( preg_match('~https://linkinghub.elsevier.com/retrieve/pii/(S[0-9]+)~i', $redirectedUrl_doi, $matches ) === 1 ) {
                 $redirectedUrl_doi = 'https://www.sciencedirect.com/science/article/pii/' . $matches[1] ;
             }
```

Reply:
20 % of DOIs warrant a special matching for Elsevier. This is a simplistic solution in case you want to only use HTTP headers. Otherwise you'd need to match the actual content of the final HTML, looking for a string like `Redirect=https%3A%2F%2Fwww.sciencedirect.com%2Fscience%2Farticle%2Fpii%2FS0167923612002370`, à la

```
preg_match('~Redirect=(https[^>]+pii%2FS0167923612002370)~', $html, $matches);
$redirectedUrl_doi = urldecode($matches[1]);
```